src/file_pipe_log/pipe_builder.rs

```diff
@@ -143,24 +206,34 @@ impl<B: FileBuilder> DualPipesBuilder<B> {
             .unwrap();
         let (append, rewrite) = pool.join(
             || {
-                Self::recover_queue(
-                    LogQueue::Append,
-                    self.file_builder.clone(),
-                    self.cfg.recovery_mode,
-                    append_concurrency,
-                    self.cfg.recovery_read_block_size.0 as usize,
-                    &self.append_files,
-                )
+                if machine_builder.buildable(LogQueue::Append) {
```

Comment:
I still don't quite like the idea of returning `Option`. How about adding another interface `replay_queue` that does single queue recovery?

Reply:
Sounds reasonable if returning `Option` is not advised in Rust. I do this just for simplicity.

Reply:
It's not specific to rust. Returning `Option` is bug-prone, we tend to unwrapping it on caller side based on assumption that could be broken at some corner cases.

Reply:
@tabokie  OK, I have update the code, Is it elegant enough about this point ?

Reply:
Looks good, https://github.com/tikv/raft-engine/pull/175#discussion_r775937403 is still unresolved. I'll do a thorough review after you address that.

Reply:
@tabokie , I have thought this for a long time, can you help to confirm if the following code is what you have refer to?


```rust
impl ReplayMachineFactory for ReplayMachineBuilder<ConsistencyChecker> {
    type Machine = ConsistencyChecker;
}

impl ReplayMachineFactory for ReplayMachineBuilder<MemTableRecoverContext> {
    type Machine = MemTableRecoverContext;
}

impl ReplayMachineFactory for ReplayMachineBuilder<TruncateMachine> {
    type Machine = TruncateMachine;

    fn new_machine(&self) -> TruncateMachine {
        let mut m = TruncateMachine::default();
        m.init(self.truncate_params.clone());
        m
    }

    fn need_truncate(&self) -> bool {
        self.truncate_params.is_some()
    }

    fn buildable(&self, queue: LogQueue) -> bool {
        return self.truncate_params.is_none()
            || self.truncate_params.as_ref().unwrap().queue.is_none()
            || self.truncate_params.as_ref().unwrap().queue.unwrap() == queue;
    }
}
```

The following code seems more elegant, But there exists error when compiling.
```rust
impl<M: ReplayMachine> ReplayMachineFactory for ReplayMachineBuilder<M> {
    type Machine = M;
}

impl ReplayMachineFactory for ReplayMachineBuilder<TruncateMachine> {
    type Machine = TruncateMachine;

    fn new_machine(&self) -> TruncateMachine {
        let mut m = TruncateMachine::default();
        m.init(self.truncate_params.clone());
        m
    }

    fn need_truncate(&self) -> bool {
        self.truncate_params.is_some()
    }

    fn buildable(&self, queue: LogQueue) -> bool {
        return self.truncate_params.is_none()
            || self.truncate_params.as_ref().unwrap().queue.is_none()
            || self.truncate_params.as_ref().unwrap().queue.unwrap() == queue;
    }
```

Reply:
Like this:
```
pub type DefaultMachineFactory<M> = PhantomData<M>;

impl<M: Default + ReplayMachine> ReplayMachineFactory for DefaultMachineFactory<M> {
  type Machine = M;
  fn new_machine(&self) -> Self::Machine { M::default() }
}

pub struct TruncateMachineFactory {
  // parameters
}

impl ReplayMachineFactory for ReplayMachineBuilder {
    type Machine = TruncateMachine;

    fn new_machine(&self) -> TruncateMachine {
        TruncateMachine::new(self.params)
    }
}
```

And remove the use of `need_truncate` and `buildable`. You can add a new interface `recover_queue` to the PipeBuilder.

Reply:
trait `need_truncate` and `buildable` have been removed and I add the corresponding method in `DualPipesBuilder`, I'am not sure if this is what you thought